restapi/src/main/java/io/stargate/web/docsapi/service/JsonSchemaHandler.java

```diff
@@ -0,0 +1,163 @@
+import com.github.fge.jsonschema.core.report.ProcessingReport;
+import com.github.fge.jsonschema.main.JsonSchemaFactory;
+import io.stargate.db.datastore.ResultSet;
+import io.stargate.db.datastore.Row;
+import io.stargate.db.query.Predicate;
+import io.stargate.db.query.builder.BuiltQuery;
+import io.stargate.db.schema.Schema;
+import io.stargate.web.docsapi.dao.DocumentDB;
+import io.stargate.web.docsapi.exception.ErrorCode;
+import io.stargate.web.docsapi.exception.ErrorCodeRuntimeException;
+import io.stargate.web.docsapi.models.JsonSchemaResponse;
+import io.stargate.web.docsapi.service.util.ImmutableKeyspaceAndTable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.inject.Inject;
+
+/** Service that handles attaching and fetching JSON schemas to/from Documents API collections. */
+public class JsonSchemaHandler {
+  private ObjectMapper mapper;
+  private final JsonSchemaFactory schemaFactory = JsonSchemaFactory.byDefault();
+  private final ConcurrentHashMap<ImmutableKeyspaceAndTable, JsonNode> schemasPerCollection =
+      new ConcurrentHashMap<>();
+  private Schema lastCheckedSchema;
+
+  @Inject
+  public JsonSchemaHandler(ObjectMapper mapper) {
+    if (mapper == null) {
+      throw new IllegalStateException("JsonSchemaHandler requires a non-null ObjectMapper");
+    }
+    this.mapper = mapper;
+  }
+
+  private void clearCacheOnSchemaChange(DocumentDB db) {
+    if (!db.schema().equals(lastCheckedSchema)) {
+      schemasPerCollection.clear();
+      this.lastCheckedSchema = db.schema();
+    }
+  }
+
+  private JsonSchemaResponse reportToResponse(JsonNode schema, ProcessingReport report) {
+    JsonSchemaResponse resp = new JsonSchemaResponse(schema);
+    report.forEach(
+        msg -> {
+          resp.addMessage(msg.getLogLevel(), msg.getMessage());
+        });
+    return resp;
+  }
+
+  public JsonSchemaResponse attachSchemaToCollection(
+      DocumentDB db, String namespace, String collection, JsonNode schema) {
+    ProcessingReport report = schemaFactory.getSyntaxValidator().validateSchema(schema);
+    JsonSchemaResponse resp = reportToResponse(schema, report);
+    if (report.isSuccess()) {
+      writeSchemaToCollection(db, namespace, collection, schema.toString());
+      ImmutableKeyspaceAndTable info =
+          ImmutableKeyspaceAndTable.builder().keyspace(namespace).table(collection).build();
+      schemasPerCollection.put(info, schema);
+      return resp;
+    } else {
+      throw new ErrorCodeRuntimeException(ErrorCode.DOCS_API_JSON_SCHEMA_INVALID);
+    }
+  }
+
+  private CompletableFuture<String> getRawJsonSchemaForCollection(
+      DocumentDB db, String namespace, String collection) {
+    BuiltQuery q =
+        db.builder()
+            .select()
+            .column("comment")
+            .from("system_schema", "tables")
+            .where("keyspace_name", Predicate.EQ, namespace)
```

Comment:
This looks risky to me. We may not know the low-level keyspace name in all cases. Could we expose comments via the main Stargate `Schema` API?

Reply:
Ah, perhaps - that would make more sense to do

Reply:
yea I forgot about CNDB here and how tenant is added into KS names 😓